jpos/src/main/java/org/jpos/iso/ISOMsg.java

```diff
@@ -285,32 +285,34 @@ public void set(String fpath, String value) {
      * @param c component
      * @throws ISOException on error
      */
-     public void set (String fpath, ISOComponent c) throws ISOException {
-         StringTokenizer st = new StringTokenizer (fpath, ".");
-         ISOMsg m = this;
-         for (;;) {
-             int fldno = parseInt(st.nextToken());
-             if (st.hasMoreTokens()) {
-                 Object obj = m.getValue(fldno);
-                 if (obj instanceof ISOMsg)
-                     m = (ISOMsg) obj;
-                 else
-                     /*
-                      * we need to go deeper, however, if the value == null then
-                      * there is nothing to do (unset) at the lower levels, so break now and save some processing.
-                      */
-                     if (c == null) {
-                         break;
-                     } else {
-                         // We have a value to set, so adding a level to hold it is sensible.
-                         m.set (m = new ISOMsg (fldno));
-                     }
-             } else {
-                 m.set (c);
-                 break;
-             }
-         }
-     }
+    public void set(String fpath, ISOComponent c) throws ISOException {
+        StringTokenizer st = new StringTokenizer (fpath, ".");
+        ISOMsg m = this;
+        for (;;) {
+            int fldno = parseInt(st.nextToken());
+            if (st.hasMoreTokens()) {
+                Object obj = m.getValue(fldno);
+                if (obj instanceof ISOMsg)
+                    m = (ISOMsg) obj;
+                else
+                    /*
+                     * we need to go deeper, however, if the value == null then
+                     * there is nothing to do (unset) at the lower levels, so break now and save some processing.
+                     */
+                    if (c == null) {
+                        break;
+                    } else {
+                        // We have a value to set, so adding a level to hold it is sensible.
+                        m.set(m = new ISOMsg(fldno));
+                    }
+            } else {
+                if (c != null)
+                    c.setFieldNumber(fldno);
```

Comment:
If  `c` is null, wouldn't be better to not add the inner fields in the first place, and just return at the beginning of the method?
The current modification is better for maintaining secondary effects for backward compatibility, but it looks weird to me.

Reply:
@alcarraz something weird happened
The change in `ISOMsg` should have been a completely separate PR from the one fixing `rc.CMF`

You are more or less right.  The semantics for `set(String, ISOComponent)` are not exactly the same as the semantics for other versions of `set()`.
For example, for a `m.set("95.2.8", (iSOComponent) null)  I would expect to be equivalent to `m.unset("95.2.8")` 

This change was specifically to fix the case of, for example, an inner `ISOAmount` that had no internal id (`fieldNumber`, or `key`) **set beforehand**, or it had one but needed to be explicited again in the fpath, and be the **same number**, otherwise the hierarchy would not be correctly created.

I tried not to break (too much!) the non-pathological usages of the method.
I'm sure that any code in existence has gone through the trouble of doing

```java
am = new ISOAmount(8);  // make the component's id explicit
m.set("95.2.8", am);    // express the whole path with 8 at the end
```

To me the semantics of `set(fpath, xxx)` should be "set xxx as the value of the last part of the path", so I shouldn't need to make the id explicit in the first line (could be -1, uninitialized, or maybe another value because the component was extracted from somewhere). Once it's inserted into the correct point, it's id should be **updated** (actually, **before** it's inserted, because that's how `m.set(c)` works.

Reply:
Sorry, I have misread the code change, I missed that you maintained the next `m.set` so I believed your intention was what I said previously, my comments have no sense whatsoever with the real change, and the comment would be another, that I will make in the new PR.

Reply:
There’s a new PR for that already, with better explanation and tests, if you want to check it out 😁